Fixes for profiling JIT code

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37587 RecordFunction in Dispatcher
* #37639 Use TensorMethods.cpp
* **#38453 Fixes for profiling JIT code**
* #37775 Memory profiling

Summary:
 - RecordFunction in JIT interpreter should exist during the execution
   of the frame, and not just when we enter the frame
 - When creating a JIT continuation in wait instruction, we'd want to
   preserve the original thread local context, right now when we resume
   execution in continuation we preserve the thread local state of the
   thread that set future value (i.e. executed a forked task)

Test Plan:
python test/test_jit.py TestJit.test_profiler
CI

Differential Revision: [D21565959](https://our.internmc.facebook.com/intern/diff/D21565959)

Commit:
[wip] Fixes for profiling JIT code

Summary:
Two fixes:
 - RecordFunction in JIT interpreter should exist during the execution
   of the frame, and not just when we enter the frame
 - When creating a JIT continuation in wait instruction, we'd want to
   preserve the original thread local context, right now when we resume
   execution in continuation we preserve the thread local state of the
   thread that set future value (i.e. executed a forked task)

Test Plan:
unittest, CI

[ghstack-poisoned]

Commit:
Update on "[wip] Fixes for profiling JIT code"


Summary:
Two fixes:
 - RecordFunction in JIT interpreter should exist during the execution
   of the frame, and not just when we enter the frame
 - When creating a JIT continuation in wait instruction, we'd want to
   preserve the original thread local context, right now when we resume
   execution in continuation we preserve the thread local state of the
   thread that set future value (i.e. executed a forked task)

Test Plan:
unittest, CI

Differential Revision: [D21565959](https://our.internmc.facebook.com/intern/diff/D21565959)

[ghstack-poisoned]

Commit:
Update on "[wip] Fixes for profiling JIT code"


Summary:
Two fixes:
 - RecordFunction in JIT interpreter should exist during the execution
   of the frame, and not just when we enter the frame
 - When creating a JIT continuation in wait instruction, we'd want to
   preserve the original thread local context, right now when we resume
   execution in continuation we preserve the thread local state of the
   thread that set future value (i.e. executed a forked task)

Test Plan:
unittest, CI

Differential Revision: [D21565959](https://our.internmc.facebook.com/intern/diff/D21565959)

[ghstack-poisoned]

Commit:
Update on "[wip] Fixes for profiling JIT code"


Summary:
Two fixes:
 - RecordFunction in JIT interpreter should exist during the execution
   of the frame, and not just when we enter the frame
 - When creating a JIT continuation in wait instruction, we'd want to
   preserve the original thread local context, right now when we resume
   execution in continuation we preserve the thread local state of the
   thread that set future value (i.e. executed a forked task)

Test Plan:
unittest, CI

Differential Revision: [D21565959](https://our.internmc.facebook.com/intern/diff/D21565959)

[ghstack-poisoned]

Commit:
Update on "[wip] Fixes for profiling JIT code"


Summary:
Two fixes:
 - RecordFunction in JIT interpreter should exist during the execution
   of the frame, and not just when we enter the frame
 - When creating a JIT continuation in wait instruction, we'd want to
   preserve the original thread local context, right now when we resume
   execution in continuation we preserve the thread local state of the
   thread that set future value (i.e. executed a forked task)

Test Plan:
unittest, CI

Differential Revision: [D21565959](https://our.internmc.facebook.com/intern/diff/D21565959)

[ghstack-poisoned]

Commit:
Update on "[wip] Fixes for profiling JIT code"


Summary:
Two fixes:
 - RecordFunction in JIT interpreter should exist during the execution
   of the frame, and not just when we enter the frame
 - When creating a JIT continuation in wait instruction, we'd want to
   preserve the original thread local context, right now when we resume
   execution in continuation we preserve the thread local state of the
   thread that set future value (i.e. executed a forked task)

Test Plan:
unittest, CI

Differential Revision: [D21565959](https://our.internmc.facebook.com/intern/diff/D21565959)

[ghstack-poisoned]

Commit:
Update on "Fixes for profiling JIT code"


Summary:
 - RecordFunction in JIT interpreter should exist during the execution
   of the frame, and not just when we enter the frame
 - When creating a JIT continuation in wait instruction, we'd want to
   preserve the original thread local context, right now when we resume
   execution in continuation we preserve the thread local state of the
   thread that set future value (i.e. executed a forked task)

Test Plan:
python test/test_jit.py TestJit.test_profiler
CI

Differential Revision: [D21565959](https://our.internmc.facebook.com/intern/diff/D21565959)

[ghstack-poisoned]

Commit:
Update on "Fixes for profiling JIT code"


Summary:
 - RecordFunction in JIT interpreter should exist during the execution
   of the frame, and not just when we enter the frame
 - When creating a JIT continuation in wait instruction, we'd want to
   preserve the original thread local context, right now when we resume
   execution in continuation we preserve the thread local state of the
   thread that set future value (i.e. executed a forked task)

Test Plan:
python test/test_jit.py TestJit.test_profiler
CI

Differential Revision: [D21565959](https://our.internmc.facebook.com/intern/diff/D21565959)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/Parallel.h:
```diff
@@ -4,7 +4,7 @@
 
 namespace at {
-namespace internal {
+namespace internal {//
```
remove?

dzhulgakov commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -943,6 +943,9 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
     static std::atomic<size_t> num_frames;
+
+    // RecordFunction object associated with this frame
+    std::shared_ptr<at::RecordFunction> record_function;
```
looks like it can be a unique_ptr (or even c10::optional)

dzhulgakov commented on torch/csrc/jit/runtime/interpreter.h:
```diff
@@ -98,8 +99,11 @@ struct InterpreterContinuation {
-      int64_t dist_autograd_context_id = 0)
-      : state(state_), stack(std::move(stack_)) {
+      int64_t dist_autograd_context_id = 0,
+      c10::optional<at::ThreadLocalState> tls_state = c10::nullopt)
```
just curious - can you remind me the reason dist_autograd_context is not part of TLS? (I think there was one but I don't recall)

ilia-cher replied:
I better refer to https://github.com/pytorch/pytorch/issues/38510 , there's been some ongoing discussion

dzhulgakov reviewed (approved):
Looks good, minor comments

Commit:
Update on "Fixes for profiling JIT code"


Summary:
 - RecordFunction in JIT interpreter should exist during the execution
   of the frame, and not just when we enter the frame
 - When creating a JIT continuation in wait instruction, we'd want to
   preserve the original thread local context, right now when we resume
   execution in continuation we preserve the thread local state of the
   thread that set future value (i.e. executed a forked task)

Test Plan:
python test/test_jit.py TestJit.test_profiler
CI

Differential Revision: [D21565959](https://our.internmc.facebook.com/intern/diff/D21565959)

[ghstack-poisoned]

Commit:
Update on "Fixes for profiling JIT code"


Summary:
 - RecordFunction in JIT interpreter should exist during the execution
   of the frame, and not just when we enter the frame
 - When creating a JIT continuation in wait instruction, we'd want to
   preserve the original thread local context, right now when we resume
   execution in continuation we preserve the thread local state of the
   thread that set future value (i.e. executed a forked task)

Test Plan:
python test/test_jit.py TestJit.test_profiler
CI

Differential Revision: [D21565959](https://our.internmc.facebook.com/intern/diff/D21565959)

[ghstack-poisoned]

Commit:
Update on "Fixes for profiling JIT code"


Summary:
 - RecordFunction in JIT interpreter should exist during the execution
   of the frame, and not just when we enter the frame
 - When creating a JIT continuation in wait instruction, we'd want to
   preserve the original thread local context, right now when we resume
   execution in continuation we preserve the thread local state of the
   thread that set future value (i.e. executed a forked task)

Test Plan:
python test/test_jit.py TestJit.test_profiler
CI

Differential Revision: [D21565959](https://our.internmc.facebook.com/intern/diff/D21565959)

[ghstack-poisoned]
